File: Toilet Thoughts/Toilet Thoughts/HomeViewController.m
//
//  HomeViewController.m
//  Toilet Thoughts
//
//  Created by Bart de Bruin on 24-11-15.
//  Copyright © 2015 BartandFouad. All rights reserved.
//

#import "HomeViewController.h"
#import "ListThoughtTableVC.h"
#import "AddThoughtVC.h"
#import <Parse/Parse.h>
#import <ParseUI/ParseUI.h>
#import "NoInternetViewController.h"
#import "Reachability.h"

@interface HomeViewController ()

@property (nonatomic) Reachability *internetReachability;


@end

@implementation HomeViewController


- (IBAction)browseThoughts:(id)sender {
    
    ListThoughtTableVC *listThoughtTableVC= [[ListThoughtTableVC alloc] initWithNibName:@"ListThoughtTableVC" bundle:nil];
    
    [UIView beginAnimations:@"View Flip" context:nil];
    [UIView setAnimationDuration:0.80];
    [UIView setAnimationCurve:UIViewAnimationCurveEaseInOut];
    
    [UIView setAnimationTransition:
     UIViewAnimationTransitionFlipFromRight
                           forView:self.navigationController.view cache:NO];
    
    
    [self.navigationController pushViewController:listThoughtTableVC animated:YES];
    [UIView commitAnimations];
    
    [self.navigationController setNavigationBarHidden:NO];
    
    [listThoughtTableVC retrieveFromParseScore];
    
}


- (IBAction)addThoughts:(id)sender {
    
    AddThoughtVC *addThoughtVC = [[AddThoughtVC alloc] initWithNibName:@"AddThoughtVC" bundle:nil];
    
    [UIView beginAnimations:@"View Flip" context:nil];
    [UIView setAnimationDuration:0.80];
    [UIView setAnimationCurve:UIViewAnimationCurveEaseInOut];
    
    [UIView setAnimationTransition:
     UIViewAnimationTransitionFlipFromLeft
                           forView:self.navigationController.view cache:NO];
    
    [self.navigationController pushViewController:addThoughtVC animated:YES];

    
     [UIView commitAnimations];
    
    [self.navigationController setNavigationBarHidden:NO];

    
}


- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    // To set the navigationbar to normal
    [self.navigationController.navigationBar setBackgroundImage:nil
                                                  forBarMetrics:UIBarMetricsDefault];
    
    NSDate *date = [NSDate date];
    NSCalendar *calender = [NSCalendar currentCalendar];
    
    NSLog(@"week: %li", (long)[[calender components: NSCalendarUnitWeekOfYear fromDate:date] weekOfYear]);
    NSLog(@"yearWhereWeekIsContainedIn: %li", (long) [[calender components:NSCalendarUnitYear fromDate:date] year]);
    NSLog(@"hourOfMoment: %li", (long)[[calender components:NSCalendarUnitHour fromDate:date] hour]);
    
    self.weekOfYear = [[calender components:NSCalendarUnitWeekOfYear fromDate:date] weekOfYear];
    self.year = [[calender components:NSCalendarUnitYear fromDate:date] year]*100;
    
    self.weekAndYear = (self.weekOfYear + self.year);
    
    if (self.weekOfYear == 1) {
        
        PFQuery *query = [PFQuery queryWithClassName:@"ToiletThought"];
        [query whereKey:@"weekNumber" equalTo:@(self.weekOfYear - 49)];
        [query orderByDescending:@"score"];
        [query findObjectsInBackgroundWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
            
            if (objects) {
                self.highestScoreObject = objects[0];
                self.highestScoringToiletThought.text = self.highestScoreObject[@"toiletThought"];
                self.highestScoringUser.text = self.highestScoreObject[@"userName"];
                
                NSNumber *highestScoreNumber = [self.highestScoreObject objectForKey:@"score"];
                self.highestScoreNumberLabel.text = [NSString stringWithFormat:@" %@", highestScoreNumber];
            }
        }];
    }
    
    
    PFQuery *query = [PFQuery queryWithClassName:@"ToiletThought"];
    [query whereKey:@"weekNumber" equalTo:@(self.weekAndYear - 1)];
    [query orderByDescending:@"score"];
    [query findObjectsInBackgroundWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
        
        if (objects) {
            self.highestScoreObject = objects[0];
            self.highestScoringToiletThought.text = self.highestScoreObject[@"toiletThought"];
            self.highestScoringUser.text = self.highestScoreObject[@"userName"];
            
            // Here we create the animation to show the highest scoring thought on the Homescreen.
            NSString *string = self.highestScoringToiletThought.text;
            NSMutableDictionary *dict = [NSMutableDictionary dictionary];
            [dict setObject:string forKey:@"string"];
            [dict setObject:@0 forKey:@"currentCount"];
            NSTimer *timer = [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(typingLabel:) userInfo:dict repeats:YES];
            [timer fire];
            
            NSNumber *highestScoreNumber = [self.highestScoreObject objectForKey:@"score"];
            int scoreInIntValue = [highestScoreNumber intValue];
            
            if (scoreInIntValue >= 0) {
                
                self.thumbsDown.hidden = YES;
                self.thumbsUp.hidden = NO;
                self.highestScoreNumberLabel.text = [NSString stringWithFormat:@" %@", highestScoreNumber];
                
                
            } else if (scoreInIntValue < 0) {
                
                self.thumbsUp.hidden = YES;
                self.thumbsDown.hidden = NO;
                self.highestScoreNumberLabel.text = [NSString stringWithFormat:@" %@", highestScoreNumber];
                
            }
        }
    }];
}

- (void)viewWillAppear:(BOOL)animated {
    
    // To set the navigationbar to normal
    [self.navigationController.navigationBar setBackgroundImage:nil
                                                  forBarMetrics:UIBarMetricsDefault];
    
    [super viewWillAppear:YES];
    
    [self.navigationController setNavigationBarHidden:YES];
    
    // No toolbar
    [self.navigationController setToolbarHidden:YES];

}


- (void)viewDidAppear:(BOOL)animated {
    
    [super viewDidAppear:YES];
    
    self.internetReachability = [Reachability reachabilityForInternetConnection];
    [self.internetReachability startNotifier];
    [self updateInterfaceWithReachability:self.internetReachability];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(reachabilityChanged:) name:kReachabilityChangedNotification object:nil];
}


- (void)reachabilityChanged:(NSNotification *)note {
    
    Reachability* curReach = [note object];
    NSParameterAssert([curReach isKindOfClass:[Reachability class]]);
    [self updateInterfaceWithReachability:curReach];
}

- (void)updateInterfaceWithReachability:(Reachability *)reachability {
    
    NetworkStatus netStatus = [reachability currentReachabilityStatus];
    if (netStatus == NotReachable) {
        
        self.thumbsDown.hidden = YES;
        self.thumbsUp.hidden = YES;
        NoInternetViewController *noInternetViewController = [[NoInternetViewController alloc] init];
        [self presentViewController:noInternetViewController animated:YES completion: nil];
        
    } else {
        
        [self dismissViewControllerAnimated:YES completion:nil];
    }
}



-(void)typingLabel:(NSTimer*)theTimer
{
    NSString *theString = [theTimer.userInfo objectForKey:@"string"];
    int currentCount = [[theTimer.userInfo objectForKey:@"currentCount"] intValue];
    currentCount ++;
    NSLog(@"%@", [theString substringToIndex:currentCount]);
    
    [theTimer.userInfo setObject:[NSNumber numberWithInt:currentCount] forKey:@"currentCount"];
    
    if (currentCount > theString.length-1) {
        [theTimer invalidate];
    }
    
    [self.highestScoringToiletThought setText:[theString substringToIndex:currentCount]];
}


@end

File: Toilet Thoughts/Toilet Thoughts/HomeViewController.swift
//
//  HomeViewController.swift
//  Toilet Thoughts
//
//  Created by Bart de Bruin on 17-12-15.
//  Copyright © 2015 BartandFouad. All rights reserved.
//

import UIKit


@objc class HomeViewController: UIViewController {
    
    var internetReachability: Reachability?
    var homeVC: UIViewController?
    var weekOfYear: NSInteger = 0
    var weekAndYear: NSInteger = 0
    var week: NSInteger = 0
    var year: NSInteger = 0
    
    @IBOutlet var highestScoringToiletThought: UILabel!
    @IBOutlet var highestScoringUser: UILabel!
    @IBOutlet var highestScoreNumberLabel: UILabel!
    @IBOutlet var thumbsUp: UIImageView!
    @IBOutlet var thumbsDown: UIImageView!
    
    var highestScoreObject: PFObject!
    
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
    
        super.init(nibName: "HomeViewController", bundle: nil)
    }

    
    
    required init?(coder aDecoder: NSCoder) {
        
        internetReachability = nil
        homeVC = nil
        highestScoreObject = nil
    
        super.init(coder: aDecoder)
    }

    
    
    @IBAction func browseThoughts(sender: AnyObject) {
        
        let listThoughtTableVC = ListThoughtTableVC()
        
        UIView.beginAnimations("View Flip", context: nil)
        UIView.setAnimationDuration(0.80)
        UIView.setAnimationCurve(.EaseInOut)
        UIView.setAnimationTransition(.FlipFromRight, forView: (self.navigationController?.view)!, cache: false)
        
        self.navigationController!.pushViewController(listThoughtTableVC, animated: true)
        UIView.commitAnimations()
        self.navigationController!.navigationBarHidden = false
        listThoughtTableVC.retrieveFromParseScore()
    }
    
    
    @IBAction func addThoughts(sender: AnyObject) {
        
        let addThoughtVC = AddThoughtVC()
        
        UIView.beginAnimations("View Flip", context: nil)
        UIView.setAnimationDuration(0.80)
        UIView.setAnimationCurve(.EaseInOut)
        UIView.setAnimationTransition(.FlipFromLeft, forView: (self.navigationController?.view)!, cache: false)
        
        self.navigationController!.pushViewController(addThoughtVC, animated: true)
        UIView.commitAnimations()
        self.navigationController!.navigationBarHidden = false
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Do any additional setup after loading the view.
        self.navigationController?.navigationBar.setBackgroundImage(nil, forBarMetrics: .Default)
        
        let date: NSDate = NSDate()
        let calender: NSCalendar = NSCalendar.currentCalendar()
        
        self.weekOfYear = calender.components(.WeekOfYear, fromDate:date).weekOfYear
        self.year = calender.components(.Year, fromDate: date).year * 100
        self.weekAndYear = (self.weekOfYear + self.year)
        
        if (self.weekOfYear == 1) {
            
            let query = PFQuery(className: "ToiletThought")
            query.whereKey("weekNumber", equalTo: self.weekOfYear - 49)
            query.orderByDescending("score")
            
            query.findObjectsInBackgroundWithBlock {
                (objects: [PFObject]?, error: NSError?)  -> Void in
                
                if (objects != nil) {
                    
                    self.highestScoreObject = objects![0]
                    self.highestScoringToiletThought.text = self.highestScoreObject["toiletThought"]as? String
                    self.highestScoringUser.text = self.highestScoreObject["userName"]as? String
                    
                    let newInt = (self.highestScoreObject ["score"]as? Int)
                    let highestScoreNumber = NSNumber.init(integer: newInt!)
                    self.highestScoreNumberLabel.text = "\(highestScoreNumber)"
                }
            }
        }
        
        let query = PFQuery(className: "ToiletThought")
        query.whereKey("weekNumber", equalTo:(self.weekAndYear - 1))
        query.orderByDescending("score")
        
        query.findObjectsInBackgroundWithBlock {
            (objects: [PFObject]?, error: NSError?)  -> Void in
            
                self.highestScoreObject = objects?[0]
                self.highestScoringToiletThought.text = self.highestScoreObject["toiletThought"]as? String
                self.highestScoringUser.text = self.highestScoreObject["userName"]as? String
                
                let newInt = (self.highestScoreObject ["score"]as? Int)
                let highestScoreNumber = NSNumber.init(integer:newInt!)
                let scoreIntValue: Int = newInt!
                if scoreIntValue >= 0 {
                    self.thumbsDown.hidden = true
                    self.thumbsUp.hidden = false
                    self.highestScoreNumberLabel.text = "\(highestScoreNumber)"
                    
                    // Here we write the animation for the homeScreen Thought
                    let string = self.highestScoringToiletThought.text
                    let dict = NSMutableDictionary()

                    dict["string"] = string
                    dict["currentCount"] = 0
                    
                    let timer = NSTimer.scheduledTimerWithTimeInterval(0.1, target: self, selector: "typingLabel:", userInfo: dict, repeats: true)
                    
                    timer.fire()
                    
                } else if scoreIntValue < 0 {
                    
                    self.thumbsDown.hidden = false
                    self.thumbsUp.hidden = true
                    self.highestScoreNumberLabel.text = "\(highestScoreNumber)"
                }
            
        }
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    
    override func viewWillAppear(animated: Bool) {
        
        super.viewWillAppear(true)
        
        self.navigationController?.navigationBar.setBackgroundImage(nil, forBarMetrics: .Default)
        self.navigationController?.setNavigationBarHidden(true, animated: true)
        self.navigationController?.setToolbarHidden(true, animated: true)
    }
    
    override func viewDidAppear(animated: Bool) {
        
        super.viewDidAppear(true)
        
        self.internetReachability = (Reachability .reachabilityForInternetConnection())
        self.internetReachability!.startNotifier()
        
        self.updateInterfaceWithReachability(self.internetReachability!)
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "reachabilityChanged:", name: kReachabilityChangedNotification, object: nil)
    }
    
    func reachabilityChanged(note: NSNotification) {
        
        let currentReachability: Reachability = note.object as! Reachability
        self.updateInterfaceWithReachability(currentReachability)
    }
    
    func updateInterfaceWithReachability(reachability:Reachability) {
        
        let netStatus: NetworkStatus = reachability.currentReachabilityStatus()
        if netStatus == NotReachable {
            
            self.thumbsDown.hidden = true
            self.thumbsUp.hidden = true
            let noInternetViewController = NoInternetViewController()
            self.presentViewController(noInternetViewController, animated: true, completion:nil)
        } else {

           self.dismissViewControllerAnimated(true, completion: nil)
        }
    }
    
    func typingLabel(theTimer:NSTimer) {
        
        let userInfoDict = theTimer.userInfo as! NSMutableDictionary
        
        let theString = (theTimer.userInfo!["string"] as! NSString)
        var currentCount = (theTimer.userInfo!["currentCount"]as! Int)
        currentCount += 1
        
        userInfoDict ["currentCount"] = currentCount
        
        if currentCount > (theString.length - 1) {
            
            theTimer.invalidate()
        }
    
        self.highestScoringToiletThought.text = theString.substringToIndex(currentCount)
        }
    }
    
   
